wandb/sdk/mailbox/wait_with_progress.py

```diff
@@ -0,0 +1,135 @@
+        The result, if it is received before the timeout.
+
+    Raises:
+        TimeoutError: If the overall timeout expires.
+        HandleAbandonedError: If the handle becomes abandoned.
+        Exception: Any exception from the display function is propagated.
+    """
+    return wait_all_with_progress(
+        [handle],
+        timeout=timeout,
+        progress_after=progress_after,
+        display_progress=display_progress,
+    )[0]
+
+
+def wait_all_with_progress(
+    handle_list: list[handles.MailboxHandle],
+    *,
+    timeout: float | None,
+    progress_after: float,
+    display_progress: Callable[[], Coroutine[Any, Any, None]],
+) -> list[pb.Result]:
+    """Wait for multiple handles, possibly displaying progress to the user.
+
+    Arguments and exceptions are the same as for `wait_with_progress`.
+
+    Returns:
+        A list where the Nth item is the result of the Nth handle.
+    """
+    if not handle_list:
+        return []
+
+    if timeout is not None and timeout <= progress_after:
+        return _wait_handles(handle_list, timeout=timeout)
+
+    start_time = time.monotonic()
+
+    try:
+        return _wait_handles(handle_list, timeout=progress_after)
+    except TimeoutError:
+        pass
+
+    async def progress_loop_with_timeout() -> list[pb.Result]:
+        with asyncio_compat.cancel_on_exit(display_progress()):
+            if timeout is not None:
+                elapsed_time = time.monotonic() - start_time
+                remaining_timeout = timeout - elapsed_time
+            else:
+                remaining_timeout = None
+
+            return await _wait_handles_async(
+                handle_list,
+                timeout=remaining_timeout,
+            )
+
+    return asyncio_compat.run(progress_loop_with_timeout)
+
+
+def _wait_handles(
+    handle_list: list[handles.MailboxHandle],
+    *,
+    timeout: float,
+) -> list[pb.Result]:
+    """Wait for multiple mailbox handles.
+
+    Returns:
+        Each handle's result, in the same order as the given handles.
+
+    Raises:
+        TimeoutError: If the overall timeout expires.
+        HandleAbandonedError: If any handle becomes abandoned.
+    """
+    results: list[pb.Result] = []
+
+    start_time = time.monotonic()
+    for handle in handle_list:
+        elapsed_time = time.monotonic() - start_time
+        remaining_timeout = timeout - elapsed_time
+        results.append(handle.wait_or(timeout=remaining_timeout))
+
+    return results
+
+
+async def _wait_handles_async(
+    handle_list: list[handles.MailboxHandle],
+    *,
+    timeout: float | None,
+) -> list[pb.Result]:
+    """Asynchronously wait for multiple mailbox handles.
+
+    Just like _wait_handles.
+    """
+    results: dict[handles.MailboxHandle, pb.Result] = {}
+
+    async def wait_single(handle: handles.MailboxHandle) -> None:
+        results[handle] = await handle.wait_async(timeout=timeout)
+
+    async with asyncio_compat.open_task_group() as task_group:
```

Comment:
do you think it makes sense to have per task group timeout (i.e. one global on top of each for each task?)

Reply:
The first timeout from a handle cancels the entire group, so it's not necessary. If I did add a group-level timeout, I would do it by starting a task that times out after `timeout`, which is what this code already does.

Reply:
yeah, that what i understood from the task group semantics, in your previous compat_async PR, but for some reason here i got confused that it won't work properly, ok, so you can ignore it.